Feat/soft delete

iWinston opened the pull request:
1. added @DeleteDateColumn
2. added the `softDelete` and `restore` methods to query builder
3. added the `softDelete` and `restore` methods to repository
4. added the `softRemove` and `recover` methods to repository
5. added the support of the cascades soft-remove and recover

iWinston commented:
Hi, I have forked TypeORM and added some features here: [TypeORM+](https://github.com/iWinston/typeorm-plus).

The README.md may be useful to show you more details of This PR.

codeandcats commented:
@iWinston thank you for this legendary effort ❤️

I hope this can be pulled in soon.

PachowStudios commented:
@iWinston if you set up a BAT wallet so I can tip you, I'll buy you a coffee. This is really great.

ksharonov commented:
Waiting for release)

pleerock commented:
haha great 😆

vinimdocarmo commented:
This will be really helpful! Thanks

hselaaron commented:
Excited for this! Much needed. 👏

mcblum commented:
I realize this is annoying and not helpful to the discussion, but is there any estimated timeline on merging this? I think there are a lot of us who are trying to figure out whether to write our own implementation / use this fork until it's merged / just wait...

Good things take time, but if there's any possibility to merge this it would be greatly appreciated!

artoodeeto commented:
@pleerock let the community review this and give you a information if its ok to merged or not, so itll be less burden to you sir. just a suggestion because I think community are waiting  for this. please :pray:

pleerock commented:
Okay since so many people want this feature, let's try to revisit the concept and merge this PR.

pleerock commented:
@iWinston good job on implementation, looks really good. Few questions:

* there is a confusion between softRemove and softDelete. Do we really need both? Also, is it the best name for this method?
* looks like you didn't implement exclude found results from find method calls. I think this is what people expect the most from this feature. We need to exclude soft deleted rows from find queries, add option to find options to enable it when needed and think if we want to exclude rows from query builder queries

Destreyf commented:
@pleerock It would be wonderful if we could get this feature merged, it would allow me to remove quite a bit of duplicated code from my builds.

I do agree that it would be beneficial (and probably could be required) to support the filtering of soft-deletes in a find query.

iWinston commented:
Thank you for your attention. @pleerock 

1. The difference between `softRemove` and `softDelete` is the same as `remove` and `delete`. One for entities, and one for given conditions.

2. I have done some work on the Query Scopes which can exclude found results from find method calls easily. And I would like to make a PR if It is considered to be a good idea.

Let me show you what the Query Scopes would like:

## Query Scopes

Query scopes allow you to add constraints to all queries for a given entity. You can register scopes in your entity:

### 1. Registering Scopes
```TypeScript
import { DeleteDateColumn } from 'typeorm-plus'

export class Entity {

    static scope = {
      'default': {
        deletedAt: IsNull()
      },
      'myScope': {
        deletedAt: Not(IsNull())
      }
    }

}
```

### 2. Applying Scopes To QueryBuilder
When you are calling `queryBuilder`, you can also apply the scope.

```TypeScript
import {createConnection} from "typeorm-plus";
import {Entity} from "./entity";

createConnection(/*...*/).then(async connection => {

    const repository = connection.getRepository(Entity);
    await repository.createQueryBuilder().setScope("myScope").getMany();

}).catch(error => console.log(error));
```
The param `scope` of the setScope function selects scope to apply to the repository. If it is false, none of the scopes will be applied. If it is undefined, the value will be "default".

### 3. Applying Scopes To Repository
When you are calling `repository`, you can apply the scope. The scope mode supports these methods of `repository`: `find`, `findOne`, `findOneOrFail`, `count`、`findByIds` And `findAndCount`.
```TypeScript
import {createConnection} from "typeorm-plus";
import {Entity} from "./entity";

createConnection(/*...*/).then(async connection => {

    const repository = connection.getRepository(Entity);
    // Delete a entity
    await repository.find({
        scope: 'myScope'
    });

}).catch(error => console.log(error));
```
The property `scope` of the find options selects scope to apply to the repository. If it is false, none of the scopes will be applied. If it is undefined, the value will be "default".


### 4. Working with Soft Delete

TypeORM's own soft delete functionality utilizes global scopes to only pull "non-deleted" entities from the database.

If the `@DeleteDateColumn` is set, the default scope will be "non-deleted".

pleerock commented:
"scopes" you are talking about is a different functionality and should come from a separate discussion and PR if we decide to go with it. Can we implement soft deletion without them? I'm sure most people just want either to load records with soft deleted rows included or records without them, most probably by default. Thanks

iWinston commented:
@pleerock 
Yes, I agree with you about loading records with soft deleted rows or records without them is enough in the most conditions.

We can just add a default "non-deleted" condition in this functionality. And maybe we also need to add the `withDeleted: true` option to `Repository` and the `withDeleted` method to `QueryBuilder`. 

But if we adds "scopes" functionality in the future,  should we change the `withDeleted: true` option and  the `withDeleted` method to another way?

pleerock commented:
> But if we adds "scopes" functionality in the future, should we change the withDeleted: true option and the withDeleted method to another way?

I don't think so. soft deletion is kinda standalone feature

> We can just add a default "non-deleted" condition in this functionality. And maybe we also need to add the withDeleted: true option to Repository and the withDeleted method to QueryBuilder.

can you implement that?

iWinston commented:
@pleerock 
Yes, I can implement this and make a PR.

goldcaddy77 commented:
Another consideration is around composite keys and soft deletion. All composite keys you add will typically want to filter out items where deletedAt is not null. Is this being considered or should it be a separate PR?

byronferguson commented:
Is the plan for this to work similar to the `paranoid: true` flag in Sequelize, which can be set at the global or table level, and the appropriate fields and where clauses are automatic for that table?

What does the current timeline look like for this feature? I would love to implement TypeORM in our refactor, but we depend on the `paranoid` flag in Sequelize currently.

mshafiq235 commented:
@pleerock Can you please estimate when would this update be available?

iWinston commented:
> @pleerock Can you please estimate when would this update be available?

Hi, I have implemented this feature, and I am adding more tests for it. 

Sorry, I'm a bit busy recently.

byronferguson commented:
Winston,

I think we all greatly appreciate your work on this feature. I look forward
to seeing the working implementation.

Scopes are something we never really leveraged in Sequelize, but I think
they could be extremely helpful.

On Thu, Jan 16, 2020 at 8:29 PM Winston Lee <notifications@github.com>
wrote:

> @pleerock <https://github.com/pleerock> Can you please estimate when
> would this update be available?
>
> Hi, I have implemented this feature, and I am adding more tests for it.
>
> Sorry, I'm a bit busy recently.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/typeorm/typeorm/pull/5034?email_source=notifications&email_token=AB6MH5ILFUU7QYXEN7B5FD3Q6EJW5A5CNFSM4JI4VXBKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJGHOMY#issuecomment-575436595>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB6MH5N25RHBEWWDGL3S3UTQ6EJW5ANCNFSM4JI4VXBA>
> .
>
-- 
Byron Ferguson
ferguson.bs@gmail.com

zholmes1 commented:
Really unfortunate that it's taken this long to merge in such a crucial feature. I know it's harsh, but I just don't consider the project complete without this. I wish I'd known this before starting my current project - otherwise I'd have just gone with the old trusty Sequelize.

goldcaddy77 commented:
> Another consideration is around composite constraints and soft deletion. All composite constraints you add will typically want to filter out items where deletedAt is not null. Is this being considered or should it be a separate PR?

@iWinston / @pleerock - have you considered how this will work with constraints?  If you use constraints with soft deletes, they'll likely all need to account for the deletedAt column

iWinston commented:
@goldcaddy77 
If the @DeleteDateColumn is set, the default global condition will be "non-deleted".

@pleerock 
Hello, I have implemented the feature of the default "non-deleted" condition and added tests for it.

I'm not sure that I should add the "non-deleted" condition to all alias or just to the main alias?

pleerock commented:
what do you guys think about `archive` and `unarchive` names? I don't really like current  `softDelete` name and `recover` name not being consistent with softDelete.

dantman commented:
@pleerock There is a little variation between ORMs (soft-delete, paranoid, etc). However the *****only***** time I ever saw the term `archive` used was to refer to a feature where instead of soft deleting a row with `deleted_at`; the ORM did an INSERT into a separate `archive` table and deleted the row in the original table.

pleerock commented:
its not necessary to use same words as other ORMs, while we are capable to use better words or introduce new words as a standard for other ORMs 🤣

goldcaddy77 commented:
I’d prefer `softDelete` and `recover` as it’s  likely somebody will want to build `archive` functionality on top of deleting (even in reading this in my gmail client I see a trash button next to an archive icon). Using the proposed naming convention would make this awkward.

pleerock commented:
lets allow people to decide if we should go with `archive and unarchive` names by 👍 or 👎 on my post above.

feather-jmalone commented:
i'd personally be fine with either set of names, but i agree with @goldcaddy77's thinking... plus, the intent/result of a method called `softDelete` -- aside from being an established convention in similar libraries -- is much clearer in name alone than `archive` would be.

pleerock commented:
okay, got you guys. Anyone has a better suggestion other than `softDelete`? In a single word 🍡

pqt commented:
> Anyone has a better suggestion other than `softDelete`? In a single word 🍡

Do you just not like the terminology @pleerock? It seems a bit late to be moving the goal posts on this now, no?

> its not necessary to use same words as other ORMs, while we are capable to use better words or introduce new words as a standard for other ORMs 🤣

It's equally not necessary to reinvent a wheel for the sake of being different, more specifically when nothing new (to a rather widely accepted term) is being introduced here.

For what it's worth, I only found this issue because I was specifically searching for TypeORM Soft Deleting. If it were something else like `trash` or nearly unrelated like `archive` I probably wouldn't have clicked the search result (unless I noticed relevant content in the meta description).

I'm 100% behind pushing standards forward but (1) nothing new is being introduced that differs from what one might expect from soft deletes if they came from another ORM and (2) the need to explain that soft deletes _exist_ but they're called something different is one of the many _cons_ to a renaming that ultimately offers no _pros_

rustamwin commented:
@pleerock what about `trash`/`restore`?

goldcaddy77 commented:
Another pro for `softDelete` and `delete` is that even for folks that use soft deleting, you’ll likely need to hard `delete` in certain situations (think GDPR).

pqt commented:
> @pleerock what about `trash`/`restore`?

I saw that one coming a mile away.

byronferguson commented:
My one concern with `softDelete` is that you'll have to have knowledge about your model, and whether it is a "paranoid" model or not. Whereas, if you call `delete` on the model and it either completes a DML `DELETE` on the table, or a "soft delete" based on the settings for the table that would be agnostic.

In the situation where you want to ensure that the record would be actually deleted regardless of the model settings you could call a `destory` method instead.

This wouldn't preclude the inclusion of a `restore` or `undelete` method on the model; however, an error should probably be thrown if that action is impossible based on the model options.

PachowStudios commented:
@byronferguson Interesting idea. I do prefer having `softDelete` be explicit though. With your proposition, you still have to have knowledge of the model to know if it's doing a soft delete or not. Now you're hiding the fact that a soft delete is happening unless you know it's a soft-deletable model.

In my own implementation of `softDelete` I override `delete` to throw an error if you try to hard delete a soft-deletable model. Then you could still have a `destroy` method to delete no matter what. Either that, or have an option like `allowHardDeleteOfSoftDeletableModel: boolean`?

byronferguson commented:
@PachowStudios our current backend suffers from model confusion where LOTS of services share a single models instance, which is probably coloring my response.

I am not opposed to the `softDelete`, but I do have the need to forcibly delete records sometimes. Providing a `destroy` method, or an option in the `delete` method that will override the "safety" of paranoid models is something that should be support. imho

diegoazh commented:
I don't believe that this great feature that community like and waits a long time is paused only because someone doesn't like names like `softDelete` and `restore`, names that other ORM's used a long time ago and all community accepts. Please land this feature that will make TypeORM more powerful and awesome 💪

pleerock commented:
Thanks @iWinston !

apollonin commented:
Amazing, thanks @iWinston !

Guys, could you please tell when are you planning to make a release with this feature?

artoodeeto commented:
@iWinston and @pleerock you are not the hero we need, but the hero we deserve. Thank you! Now ill wait for the documentation on how to use it. Thank you again!

pleerock commented:
> Guys, could you please tell when are you planning to make a release with this feature?

I'll release a new version on Friday

lfmmachado commented:
@iWinston @pleerock Awesome work guys, congrats!!!

artoodeeto commented:
@pleerock  Hello, sir can you also add documentation for cascading, nothing in there. If you want help just let me know. thank you. :smile:

vectrobyte commented:
Waiting for the documentation about soft deletes❤️

SkvokeN commented:
Hello, thank you for your work. I have issue with soft remove.

`
@Entity({ name: 'user' })
class UserEntity {
   @PrimaryGeneratedColumn()
   public id: number;

  @ManyToMany(
    () => LikeEntity,
    {cascade: true}
  )
  @JoinTable({
    name: 'user_like',
    joinColumn: {
      name: 'like_id',
      referencedColumnName: 'id',
    },
    inverseJoinColumn: {
      name: 'user_id',
      referencedColumnName: 'id',
    },
  })
  public functionalTypes?: [LikeEntity];
}

@Entity({ name: 'like' })
class LikeEntity {
   @PrimaryGeneratedColumn()
   public id: number;
  @DeleteDateColumn({name: 'deleted_at'})
  public deletedAt?: Date;
}
`
`
 likeRepository.softRemove(like);
`

but after soft remove row in many to many table still exists

felixle236 commented:
Awesome guys, this feature is very useful.
So...can we have a function like `onlyDeleted()` to get the data has been deleted only?

johannesschobel commented:
Dear @SkvokeN , there is a new `cascade` param, which is called `soft-delete` instead of `delete` or something. Try using this one to soft-delete over relationships!

artoodeeto commented:
@johannesschobel do you have it on the documentation sir? I think some of the newcomers are waiting for the documentation. I was just reading and exploring the code base which is cumbersome.

johannesschobel commented:
Dear @artoodeeto ,

no, i don't think that it has been added to the docs yet. At least I was not able to find it yesterday. Note, that this feature has not even been documented in the `CHANGELOG.md` file of this project ;)

The new `RelationOptions` look like this

https://github.com/typeorm/typeorm/blob/17f2fff359f8626d0593d6ef4b659593e9fb60e9/src/decorator/options/RelationOptions.ts#L8-L17

I think, you should use, for example, `cascade: 'soft-remove'`, if a relationship should be soft-deleted if the "parent" is (soft)-deleted.

Please note that **i have not used / tested this feature** for now, haha

Hope this helps,
Johannes

artoodeeto commented:
Hello @johannesschobel Thank you for your awesomeness response. Anyway sir I have an [issue](https://github.com/typeorm/typeorm/issues/5660)  maybe you could help me. This is unrelated but I think you could help me with this one. Thank you.

iWinston commented:
Hi, I think it will be helpful to show you some doc copied from my fork [repo](https://github.com/iWinston/typeorm-plus) :
## Soft Deleting

### 1. Including Soft Deleted Entities
In addition to actually removing records from your database, TypeORM+ supports "soft delete". When entities are soft deleted, they are not actually removed from your database. Instead, an attribute that records the delete time is set on the entity and inserted into the database. If the attribute is a non-null value, the entity has been soft deleted. To enable soft deletes for an entity, use the `@DeleteDateColumn` on the entity:

```TypeScript
import { DeleteDateColumn } from 'typeorm-plus'

export class Entity {

    @DeleteDateColumn({ name: 'deleted_at' })
    public deletedAt: Date

}
```


### 2. Applying Soft Delete To QueryBuilder

`@DeleteDateColumn` is a special column that is automatically set to the entity's delete time each time you call soft-delete of entity manager or repository. You don't need to set this column - it will be automatically set.

```TypeScript
import {createConnection} from "typeorm-plus";
import {Entity} from "./entity";

createConnection(/*...*/).then(async connection => {

    await connection
      .getRepository(Entity)
      .createQueryBuilder()
      .softDelete()

    // And You can restore it using restore;
    await connection
      .getRepository(Entity)
      .createQueryBuilder()
      .restore()

}).catch(error => console.log(error));
```


### 3. Applying Soft Delete To Repository

```TypeScript
import {createConnection} from "typeorm-plus";
import {Entity} from "./entity";

createConnection(/*...*/).then(async connection => {

    const repository = connection.getRepository(Entity);

    // Delete a entity
    await repository.softDelete(1);

    // And You can restore it using restore;
    await repository.restore(1);

    // Or You can soft-delete them using softRemove
    const entities = await repository.find();
    const entitiesAfterSoftRemove = await repository.softRemove(entities);

    // And You can recover them using recover;
    await repository.recover(entitiesAfterSoftRemove);

}).catch(error => console.log(error));
```

### 4. Cascading Soft Deletes

This example show what the cascading soft deletes behaves in TypeORM+.

```TypeScript
const category1 = new Category();
category1.name = "animals";

const category2 = new Category();
category2.name = "zoo";

const question = new Question();
question.categories = [category1, category2];
const newQuestion =  await connection.manager.save(question);

await connection.manager.softRemove(newQuestion);
```
As you can see in this example we did not call `save` or `softRemove` for category1 and category2. But They will be automatically saved and soft-deleted when the `cascade` of relation options is set to true like this:

```typescript
import {Entity, PrimaryGeneratedColumn, Column, ManyToMany, JoinTable} from "typeorm-plus";
import {Category} from "./Category";

@Entity()
export class Question {

    @PrimaryGeneratedColumn()
    id: number;

    @ManyToMany(type => Category, category => category.questions, {
        cascade: true
    })
    @JoinTable()
    categories: Category[];

}
```
## Default global condition

TypeORM's own soft delete functionality utilizes global condition to only pull "non-deleted" entities from the database.

If the `@DeleteDateColumn` is set, the default scope will be "non-deleted".

artoodeeto commented:
@iWinston thank you. Can I make a PR for this to put in main typeorm documentation or are you doing it?

artoodeeto commented:
@iWinston hello again sir. sorry to bother you but may I asked why is it `softRemove` method not implemented in `BaseEntity`? Thank you and keep rockin.

avin-kavish commented:
@iWinston do we really need TypeORM+ can't we work on bringing those features to this instead, we wouldn't want to split the community (unless there's a compelling reason) especially after all the hard work that the author has put into this. (I read somewhere that he built this full time over 3 years.)

johannesschobel commented:
@avin-kavish the softdelete feature has been merged into this project. Everything is fine

feather-jmalone commented:
@avin-kavish i think the compelling reason -- and I truly mean no disrespect or ingratitude towards @pleerock, @Kononnable, or any other contributor -- is that typeorm's roadmap / release cycle has had its ups and downs since the beginning; 'soft delete' in particular was first proposed in june 2017, with significant interest from the community (https://github.com/typeorm/typeorm/issues/534), but it never took off.

coupled with the fact that this forum is no stranger to cynical discussions about the state of the project overall, i wouldn't be surprised if @iWinston simply wanted to build some features on his own timeline.

but as @johannesschobel said, the work is now merged into the main project, and to me, the end result of this thread has been a very positive signal that typeorm is alive and well (and listening to its community)

avin-kavish commented:
@johannesschobel I know this one is merged, but there's more plus features in the other one.

Suggestion: Instead of TypeORM+, being a fork, can't we have it so that it will augment this library when imported? (I don't know if it's possible though) @iWinston

iWinston commented:
@avin-kavish Hi, I have tried to patch TypeORM to add extra functionalities : [TypeORM-patch](https://github.com/iWinston/typeorm-patch), but it's too complex to finish it.

Sorry, maybe I shouldn't name it as TypeORM+, I just attend to add more functionalities for our own team at the beginning. 

As more and more developers noticed the repo, I kept the repo just intended to get advice from them easily. And when the functionalities became mature, I  made some PRs for TypeORM. I have mentioned this on [1 Oct 2019](https://github.com/iWinston/typeorm-plus/issues/1#issuecomment-536927201).

Thanks for @pleerock and @Kononnable, TypeORM is a wonderful repo and very helpful for me.

iWinston commented:
> @iWinston hello again sir. sorry to bother you but may I asked why is it `softRemove` method not implemented in `BaseEntity`? Thank you and keep rockin.



> BaseEntity

Thank you to point it out! I missed this part, maybe you can make a PR for it.

iWinston commented:
> @iWinston thank you. Can I make a PR for this to put in main typeorm documentation or are you doing it?

Hi, I will make a PR for the doc if it's necessary.

PachowStudios commented:
Adding an extensibility framework to TypeORM would be really great. That way we can start bringing new features at our own pace without resorting to forks.

Thanks again for all your work @iWinston

artoodeeto commented:
@iWinston Hello sir. for sure I can create a PR for this. Can I use your repo though so you can guide me? this is embarrassing but I'm not really sure about what's happening in the code. :sad: I can also make a pr for the docs this weekend if I'm not busy.

TheDuckman commented:
So, I have a question. Are the soft-deleted entities supposed to show up in the related entities when I execute a findOne()? Is there a specific way to suppress the soft-deleted related entities? I'm looking for a way of doing this using the repository, not the queryBuilder, if possible.

The BloodPressure entities have the "deleted" field correctly set by TypeORM when I call softDelete() but they still show up in the array of related entities.

Can someone help me out with this? Thanks in advance! (sample code below)

	@Entity({name: "appointment"})
	export class Appointment {
		(...)
		
		@OneToMany(type => BloodPressure, blood_pressure_measurements => blood_pressure_measurements.appointment, {
			eager: true,
			cascade: ['insert', 'update'],
		})
		blood_pressure_measurements?: BloodPressure[];
	}
	
	@Entity({name: "blood_pressure"})
	export class BloodPressure {
		(...)
		
		@DeleteDateColumn({ name: 'deleted' })
		deleted?: Date
		
		@ManyToOne(type => Appointment)
		@JoinColumn({ name: 'appointment_id' })
		appointment?: Appointment;
	}
	
	const repo = getManager().getRepository(Appointment);
	const data = await repo.findOne({ id: appointmentId });

iWinston commented:
@TheDuckman 
The default condition of "non-deleted" is only applied to the main alisa.

When you don't want to suppress the soft-deleted related entities in your left join, maybe you should add the condition to the JOIN's ON clause.
[Soft delete field causes left join not to return my “left” side data?](https://dba.stackexchange.com/questions/176654/soft-delete-field-causes-left-join-not-to-return-my-left-side-data)

TheDuckman commented:
But I don't see a way of setting the "ON" clause explicitly on the JoinOptions of FindOneOptions (repository findOne() options types). From what I've seen, that is only possible when using the queryBuilder. Am I missing something?

If that's the only way, I'll have to use the queryBuilder but I'd really rather use the repository.

ruslanguns commented:
I am not sure if opening an issue that is why I am writting here... 

I just tried to make cascade working with soft-delete but is not working with child related tables. Only works with first entity... 

This is the parent entity()
```ts
@Entity('user')
export class User {
  @PrimaryGeneratedColumn({ type: 'bigint' })
  id: number;

  @Index({ unique: true })
  @Column({ type: 'varchar', length: 20, nullable: false })
  username: string;

  @Column({ type: 'varchar', length: 500, nullable: true, select: false })
  password: string;

  @OneToMany(type => UserMetadata, meta => meta.user, { eager: true, cascade: true })
  @JoinTable()
  metadata: UserMetadata[];

  @DeleteDateColumn({ name: 'deleted_at' })
  public deletedAt: Date;
}
````

The user metadata entity looks like that:

```ts
@Entity('user_metadata')
export class UserMetadata {
  @PrimaryGeneratedColumn({ type: 'bigint' })
  id: number;

  @Column({ type: 'varchar', length: 255, nullable: true })
  key: string; 

  @Column({ type: 'varchar', length: 255, nullable: true })
  value: string;  

  @Index()
  @ManyToOne(type => User, (user) => user.metadata)
  @JoinColumn({ name: 'user_id' })
  user!: User;

  @DeleteDateColumn({ name: 'deleted_at' })
  public deletedAt: Date;
}
```

iulian03 commented:
@ruslanguns To recover a soft-deleted entity with all of it's relations, you should provide "withDeleted: true" as an Option when getting that specific entity:

`let options: FindOneOptions<YourEntity> = {
            relations: [
                'rel1',
                'rel2'
            ],
            withDeleted: true
        };

        return await this.yourRepository.findOne(id, options);`

PS: Also provide what relations you want to be fetched with your Entity.

jaequery commented:
does this not work for ActiveRecord at the moment?

jaequery commented:
> > @iWinston hello again sir. sorry to bother you but may I asked why is it `softRemove` method not implemented in `BaseEntity`? Thank you and keep rockin.
> 
> > BaseEntity
> 
> Thank you to point it out! I missed this part, maybe you can make a PR for it.

is there anyone working on this?

iWinston commented:
@jaequery Hi, I can make a PR for it in the next few days.

jaequery commented:
@iWinston that would be amazing Winston, thanks

mertinop commented:
Hello,
is [the documentation](https://github.com/typeorm/typeorm/pull/5034#issuecomment-597972079) found on this PR in the official docs?

jaequery commented:
Hi @iWinston was this ever implemented?

johannesschobel commented:
See the "merged" flag on the PR..

dan-j commented:
@iWinston is there any limitations on specific databases for this to work? More specifically, for my use case we need this functionality in MongoDB?

From my small understanding of how TypeORM works, I believe the implementation of the MongoDB driver is quite different under the hood compared to SQL databases, a cmd+f in this thread doesn't seem to mention it 🧐

danielzev commented:
@iWinston what are your thoughts on respecting the `@DeleteDateColumn` in the `eager loaded` sub-entities, like in @TheDuckman's example with his `BloodPressure` entity in https://github.com/typeorm/typeorm/pull/5034#issuecomment-600080106

In his example, the `Appointment` entity would contain an array of `BloodPressure` objects, though it would return even soft-deleted ones. 

YES, we could use the query builder and build our own queries everywhere throughout the app. Though if that's the case, then we don't really need the `@DeleteDateColumn`, and probably not a good idea to use it if some spots use it and some spots don't.

I feel like everywhere throughout the ORM, and it's internal query builders, it should respect the `field IS NULL` where statement. Or maybe it's an option you can pass through in the repository functions. Something like `cascadeDeleteCheck` or something like that.

Ihor-Lavrov commented:
@pleerock @iWinston hi guys seems like soft deletion and `withDeleted` flags work perfect, but I've wasted a ton of time to find a description of `withDeleted` flags in this thread. There are now additional info anywhere in docs or somewhere else. Could you please update the description of soft deletion and approaches to work within soft deletion in typeorm docs included with `withDeleted` flag. It would be very helpful! Thanks

Ihor-Lavrov commented:
@pleerock @iWinston  
And I've found a possible issue: if we have subqueris seems like we should add `withDeleted` methods both to main query and subquery. As for me it's little bit weired and may cause to overhead conditional comparing. 
`repository
            .createQueryBuilder('entity')
            .innerJoin(
                subQuery => {
                    subQuery
                        .select('search.pk', 'pk')
                        .withDeleted()
                        .from(entity, 'search');
                    return subQuery;
                },
                'sub',
                'sub.pk=entity.pk'
            )
            .withDeleted()
            .addOrderBy(`entity.pk`, 'DESC');
`
(little bit cut code fragment) 
I'd appreciate if you may comment this.

nolazybits commented:
@danielzev have you found any solution. We hit the same problem where having relationship with eager loading will load them all even the soft deleted ones.
